src/hypothesis/internal/conjecture/floats.py

```diff
@@ -0,0 +1,237 @@
+which has the following properties:
+
+1. NaNs are ordered after everything else.
+2. Infinity is ordered after every finite number.
+3. The sign is ignored unless two floating point numbers are identical in
+   absolute magnitude. In that case, the positive is ordered before the
+   negative.
+4. Positive floating point numbers are ordered first by int(x) where
+   encoding(x) < encoding(y) if int(x) < int(y).
+5. If int(x) == int(y) then x and y are sorted towards lower denominators of
+   their fractional parts.
+
+
+Unfortunately it's not quite so simple. There is a complication around zero,
+which is that we want to treat 0 as if it were a normal integer, but its
+representation in IEEE floating point gives it a high negative exponent. So
+we then have a separate layer on top of our encoding for handling zero.
+
+The format of our encoding of floating point goes as follows:
+
+    [exponent] [mantissa] [sign]
+
+Each of these is the same size their equivalent in IEEE floating point, but are
+in a different format.
+
+We translate exponents as follows:
+
+    1. The maximum exponent (2 ** 11 - 1) is left unchanged.
+    2. We reorder the remaining exponents so that all of the positive exponents
+       are first, in increasing order, followed by all of the negative
+       exponents in decreasing order (where positive/negative is done by the
+       unbiased exponent e - 1023).
+
+We translate the mantissa as follows:
+
+    1. If the unbiased exponent is <= 0 we reverse it bitwise.
+    2. If the unbiased exponent is >= 52 we leave it alone.
+    3. If the unbiased exponent is in the range [1, 51] then we reverse the
+       low k bits, where k is 52 - unbiased exponen.
+
+The low bits correspond to the fractional part of the floating point number.
+Reversing it bitwise means that we try to minimize the low bits, which kills
+off the higher powers of 2 in the fraction first.
+"""
+
+
+MAX_EXPONENT = 0x7ff
+
+SPECIAL_EXPONENTS = (0, MAX_EXPONENT)
+
+BIAS = 1023
+MAX_POSITIVE_EXPONENT = (MAX_EXPONENT - 1 - BIAS)
+
+
+def exponent_key(e):
+    if e == MAX_EXPONENT:
+        return float('inf')
+    unbiased = e - BIAS
+    if unbiased < 0:
+        return 10000 - unbiased
+    else:
+        return unbiased
+
+    return -e
+
+
+ENCODING_TABLE = array('H', sorted(hrange(MAX_EXPONENT + 1), key=exponent_key))
+DECODING_TABLE = array('H', [0]) * len(ENCODING_TABLE)
+
+for i, b in enumerate(ENCODING_TABLE):
+    DECODING_TABLE[b] = i
+
+del i, b
+
+
+def decode_exponent(e):
+    """Take draw_bits(11) and turn it into a suitable floating point exponent
+    such that lexicographically simpler leads to simpler floats."""
+    assert 0 <= e <= MAX_EXPONENT
+    return ENCODING_TABLE[e]
+
+
+def encode_exponent(e):
+    """Take a floating point exponent and turn it back into the equivalent
+    result from conjecture."""
+    assert 0 <= e <= MAX_EXPONENT
+    return DECODING_TABLE[e]
+
+
+REVERSE_BITS_TABLE = bytearray([
```

Comment:
How is the magic table generated?  I'd like explanatory comments at a minimum, and if possible generate the actual bytes at runtime (with explanatory comments).

Reply:
I err may have copied it off the internet. I'll add a comment and a test to verify that it's correct (but would prefer not to generate it at load time because it's kinda tedious to do efficiently)